### PR TITLE
[VecOps,RVec] Proposal for sorting utils

### PR DESCRIPTION
**Pull request by stwunsch**
Making a proposal to add utils for sorting of collections. The use-case is to get the indices of the largset/smallest elements of a vector, e.g., `Muon_pt` and retrieve a subset of other vectors from these, e.g., `Muon_eta`, ... . Already solved by @amadio in [this](https://root-forum.cern.ch/t/sort-struct-of-array-format-with-rvec/29485) forum post.

```cpp
root [1] using namespace ROOT::VecOps;
root [2] RVec<float> x({3, 1, 2})
(ROOT::VecOps::RVec<float> &) { 3, 1, 2 }
root [3] auto i = Argsort(x)
(ROOT::VecOps::RVec<unsigned long> &) { 1, 2, 0 }
root [4] auto y = ByIndices(x, i)
(ROOT::VecOps::RVec<float> &) { 1, 2, 3 }
```

Functionalities and naming TBD.

### CODE REVIEW

File: math/vecops/inc/ROOT/RVec.hxx [hunk 1]
```diff
@@ -667,6 +667,27 @@ void swap(RVec<T> &lhs, RVec<T> &rhs)
    lhs.swap(rhs);
 }
 
+/// Return indices that sort the vector
+template <typename T>
+RVec<size_t> Argsort(const RVec<T> &v)
+{
+   using size_type = typename std::vector<T>::size_type;
+   RVec<size_type> i(v.size());
```
Comment (amadio) [inline math/vecops/inc/ROOT/RVec.hxx]: Return is `RVec<size_t>` while this is `RVec<size_type>`. Maybe it's better to keep it consistent by using the same type in both places.

File: math/vecops/inc/ROOT/RVec.hxx [hunk 2]
```diff
@@ -667,6 +667,27 @@ void swap(RVec<T> &lhs, RVec<T> &rhs)
...
+   return i;
+}
+
+/// Return elements of a vector at given indices
+template <typename T>
+RVec<T> ByIndices(const RVec<T> &v, const RVec<typename std::vector<T>::size_type> &i)
+{
+   RVec<T> r(i.size());
+   for (unsigned int k = 0; k < i.size(); k++)
+      r[k] = v[i[k]];
```
Comment (amadio) [inline math/vecops/inc/ROOT/RVec.hxx]: Why not a range-based for loop, as in [my post](https://root-forum.cern.ch/t/sort-struct-of-array-format-with-rvec/29485/5)?

Comment (stwunsch) [reply math/vecops/inc/ROOT/RVec.hxx]: I don't know, does it make any difference performance wise? But I agree that it looks much nicer.

File: math/vecops/inc/ROOT/RVec.hxx (hunk 1 above)
Comment (stwunsch) [reply math/vecops/inc/ROOT/RVec.hxx]: ah right, forgot the `size_t` here from my proof of concept implementation

**Comment by stwunsch** [comment]
@rcaspart Does this fit your problem with object selection from vectors?

**Comment by rcaspart** [comment]
Hi Stefan,
thanks a lot for the work on providing these helpers directly in ROOT.
Indeed these fit the use-cases I have in mind very nicely.

**Comment by stwunsch** [comment]
@phsft-bot build

**Pull request merged**